### DISCUSSION TITLE
ブランチ戦略とは？

### DISCUSSION THREAD

**Discussion by kagomen** [lgtm-factory/lgtm-factory#14] [Scrap]
## 関連イシュー
#12

**Comment by kagomen** [comment]
## まず、ブランチとは

- Git上で別々の作業を並行して行うための仕組みのこと
- 一人の時はいつもmainブランチにcommit/pushしてるけど、共同開発ではブランチを作成して、自分だけの作業場を確保して、そこで作業をする
- 作業が終わったら、プルリクを出してコードレビューをしてもらう
- マージされたら、やっとmainブランチに変更が反映される！

## ブランチ戦略とは

> ブランチは誰でもいくつでも作ることができる。ブランチのブランチ、とかもできちゃう。無法地帯の予感・・・！！！というわけで、無法地帯にならないように計画的にブランチを作って運用していくために編み出されたもの、それがブランチ戦略。

- [参照](https://qiita.com/trsn_si/items/cfecbf7dff20c64628ea#2%E3%83%96%E3%83%A9%E3%83%B3%E3%83%81%E6%88%A6%E7%95%A5%E3%81%A8%E3%81%AF)

## 代表的なブランチ戦略の種類

- git-flow, GitHub Flow, gitworkflows, 安定trunkパターン, メインラインモデル, GitLab Flow

- [参照](https://qiita.com/trsn_si/items/cfecbf7dff20c64628ea#3%E4%BB%A3%E8%A1%A8%E7%9A%84%E3%81%AA%E3%83%96%E3%83%A9%E3%83%B3%E3%83%81%E6%88%A6%E7%95%A5)

色々あるなー

**Reply by kagomen** [reply]
## 小規模開発ではどのブランチ戦略を採る？

### GitHub Flow

が、シンプルで良さそう

<img width="600" alt="20230802093956" src="https://github.com/user-attachments/assets/4d5e20a9-a9f6-4ce9-a45b-2f97da0af2af">

> なぜ我々のチームがGitHub Flowを採用したかというと、Insight Edgeでは小回りの効く比較的小規模なチームで素早く開発をすることが多かったから、が理由になります。Git Flowのような厳密な管理をするブランチ戦略は、先の説明の通りリリースまでの工程や確認が多くなります。そのため、こまめに機能をリリースしながらシステムをアップデートしていくというアジャイル的な我々の開発スタイルにミートしませんでした。GitHub Flowのようにシンプルでこまめにリリースを重ねていける開発スタイルが小規模チームは適していると考えました。

### GitHub Flowのデメリット

> GitHub Flowはmainブランチの内容が常にリリースとなることから、検証環境へのリリースやリリースバージョン管理が出来ないというデメリットがあります。

- [参照](https://techblog.insightedge.jp/entry/branch-strategy-github-flow)

**Reply by kagomen** [reply]
[上記の参照先](https://techblog.insightedge.jp/entry/branch-strategy-github-flow)に開発手順として、

> 1. mainブランチからfeatureブランチを作成
> 1. featureブランチで開発完了後、mainブランチへPull Request
> 1. featureブランチの内容をレビュー。問題なければmainブランチにリポジトリ管理者がマージ
> 1. mainブランチにマージ後、stg環境へリリース
> 1. stg環境にてマージされた内容の動作を確認
> 1. stg環境で問題なければリポジトリ管理者がリリースバージョンを示すtagを付与
> （前回の本番環境へのリリース内容と今回のリリース内容の差分を示すリリースノートを作成）
> 1. tagの付与されたコミットをprd環境へリリース

とあるけど、（今回は5以降はやらないにしても、）「stg環境へリリース」「prod環境へリリース」って、実際どういう操作が行われるんだろう？
ホスティングサービスのサイトの、ダッシュボードで「リリース」をする感じ・・・？
あと、リリースノートを作成してからリリースできるんだ・・・？

**Comment by kagomen** [comment]
## ブランチの命名規則

> `feature/xxxx`
> mainブランチから切ります。開発する機能に応じて作成します。xxxxには開発する機能の内容が書かれます。

例えば、`feature/to-do-list`とか、`feature/contact-form`　みたいな感じかな

## 個人的に思うこと

結構feature（機能名）をつけるの難しい気がする、イシューを分割すればするほど・・・
なので、 `feature/#12`　（#12はイシュー番号）でも良い気がする
どう思いますか・・・？ わかりにくいかな？

あとは、`feature` の他に `fix` は必要かも？
新たに追加する時は`feature`、ソースコードを修正する時は`fix`


## 追記

それとも、参照元の記事の企業では、contact-formに関係のあるものはすべて`feature/contact-form`にするってことなのかな？
ブランチ名かぶりそう、と思ったけど、イシュー番号を後ろにつけたら（`feature/contact-form-#12`）問題ないか

**Reply by kagomen** [reply]
@kazzyfrog どう思いますか？👀

**Reply by kazzyfrog** [reply]
`feature/xxxx`
に関しては、`feature `は、
作業に応じたprefixを付けるということかと思います！

なので、
「`これから行う作業のラベル/簡潔な作業内容-#12`」
とかが、一般的ですかね！

機能追加じゃなかったら、`feature`とは、別のprefixになる感じ、、ですかね！

**Reply by kazzyfrog** [reply]
ただ、基本的に、`feature/`ブランチを切って、
その中で、レビューなどからで、適宜`fix`とかのコミットを積んで、マージされて、タスクが完了する感じですかね！

なんというか、しっかり機能追加のまとまりを完了させてからマージする感じ。。！？

**Reply by kagomen** [reply]
@kazzyfrog 

すみません！自分の質問が雑すぎて多分内容が伝わってないと思うので、もう一度ご意見ください！🙏

参照元の記事の企業では、
例えば、作業内容がコンタクトフォームのリファクタリングだった場合、`refactor/contact-form-#12`とするのではなく、`feature/contact-form-#12`としてるんじゃないか？と、私が勝手に推測しており、
その推測のもと、上記のやり方についてどう思いますか？という質問でした・・・！

作業**内容**に注目するのではなく、作業**対象**に注目して、`feature/XXX`というブランチを切ってるんじゃないか？ということです！（記事内ではmainブランチとfeatureブランチ以外無さそうな感じだったので・・・！）

そして私は、これに対して、わりと良いかも！と思ってます！
ファスコンでも結構prefixどうするか問題があったと思うんですが、こういう切り口で`feature`に統一してしまえば、かなり楽なんじゃないかな・・・？と思った次第です

先ほど↑で、

> feature の他に fix は必要かも？

と書いたのですが、`fix`すら必要ないんじゃないかな？と考えてます
理由としてはhondさんがおっしゃる通り、細かな作業内容についてはコミットメッセージの領分なのかなーと思ったからです

なので、**ブランチ＝作業場**と考え、prefixをfeatureに全統一しても良いのでは！？と現在は考えています

あとは個人的に一番楽なのは`feature/#12`という形式かなー、とも思うのですが、意味的に`機能#12`って変なので、もしこの形式にするのであれば、prefixは`feature`ではないものに統一したいです
って感じです！

prefixを詳細に分けた方が良い理由などがもしあれば、教えていただきたいです！

長くなってすみません💦

**Reply by kazzyfrog** [reply]
> feature/contact-form-#12としてるんじゃないか？と、私が勝手に推測しており、
その推測のもと、上記のやり方についてどう思いますか？という質問でした・・・！
~
作業内容に注目するのではなく、作業対象に注目

理解しました！👍

> prefixをfeatureに全統一

個人的には、もし統一する場合は、
そもそもprefixは、いらないのでは？と思いました！

そもそも、「ブランチ名＝作業場」という認識は同じで、
ブランチ名の命名規則のないチームの場合は、単純に作業内容に応じたブランチ名をつけるかと思います！
例：`contact-form-#12`とか、`add-contact-form`とか！

ただ、それを少し分かりやすくするための、ある種の工夫？として、
タグ付け的に、先頭にprefixを付ける慣習がある感じですかね！

なので、prefixなしでも自分的にはOKですよ！
ただ、prefix を featureだけというのは、なんかprefixの意味がないかな？と、少し思いました！
ドキュメント関連の作業でもfeatureだと、逆に混乱しそうなので！😎

**Reply by kagomen** [reply]
@kazzyfrog 

> ブランチ名の命名規則のないチームの場合は、単純に作業内容に応じたブランチ名をつけるかと思います！ただ、それを少し分かりやすくするための、ある種の工夫？として、タグ付け的に、先頭にprefixを付ける慣習がある感じですかね！

なるほど！そうだったのですね！

> prefix を featureだけというのは、なんかprefixの意味がないかな？と、少し思いました！ドキュメント関連の作業でもfeatureだと、逆に混乱しそうなので！😎

たしかに、そうですよね……。

> prefixなしでも自分的にはOKですよ！

ありがとうございます🙏
もう一度、色んな方向から考えてみます・・・！

**Comment by kagomen** [comment]
> [GitHubさえあればブランチ名は適当でいいのかもしれない](https://windyakin.hateblo.jp/entry/2021/03/04/232006)

> GitHub上で行われるPull Requestを通した開発サイクルの中で、ブランチ名が評価されるタイミングがまったくないことに気づいた。だからといって流石にブランチ名に変なことを書くのははばかられるが、おそらくブランチ名にどんなに正しい英語が書かれていてもプラスの材料にはならないし、逆に支離滅裂な英語だとしても特に問題が起きるとも思えない。そう例え真心込めてブランチ名手打ちしたとしても、機械的に生成した文字列でブランチ名を決めたとしても、変更内容とPull Requestの内容が一緒であればそれは一緒の変更である。

実際に機械的にブランチ名を決めることを実践した結果、デメリットとしては、

> シェルの画面上でPull Requestの一覧を表示したり、Pull RequestのIDでチェックアウトしたりできるのでブランチ名で判断できない

というのがあったみたい

**Reply by kagomen** [reply]
とするなら、単に`#12`的な命名は良くなさそうだということはわかる

**Reply by kagomen** [reply]
## GitHubのブランチ名がユーザーに表示されるタイミング

GitHubのブランチ名は、様々な場面でユーザーの目に触れます。gitコマンド実行時だけでなく、GitHubのWebインターフェース上でも確認できます。

### gitコマンド実行時

* **git status:** 現在のブランチ名を表示します。
* **git branch:** 全てのローカルブランチの一覧を表示します。
* **git checkout:** ブランチを切り替える際に、切り替える先のブランチ名を指定します。
* **git merge:** 別のブランチの変更をマージする際に、マージ元のブランチ名を指定します。
* **git push:** リモートリポジトリにブランチをプッシュする際に、プッシュするブランチ名を指定します。
* **git pull:** リモートリポジトリから変更をフェッチし、マージする際に、マージするブランチ名を指定します。

### GitHubのWebインターフェース

* **リポジトリのトップページ:** 現在のデフォルトブランチ名が表示されます。
* **コードタブ:** ファイル一覧の上部にブランチ選択のドロップダウンメニューがあり、全てのブランチを確認できます。
* **プルリクエスト:** ベースブランチとヘッドブランチが表示されます。
* **コミット履歴:** 各コミットがどのブランチで行われたかが表示されます。

### その他

* **IDEやコードエディタ:** 使用しているIDEやコードエディタによっては、プロジェクトを開いた際に現在のブランチ名を表示したり、ブランチを切り替えるための機能を提供しているものがあります。
* **CI/CDパイプライン:** CI/CDパイプラインの設定ファイルでは、どのブランチに対してビルドやテストを実行するかを指定するため、ブランチ名が記述されています。

### ブランチ名が表示される理由

ブランチ名は、開発中のソフトウェアの異なるバージョンを管理するための重要な情報です。どの機能がどのブランチで開発されているのか、どのバージョンのコードが本番環境にデプロイされているのかなどを把握するために、ブランチ名が頻繁に参照されます。

**まとめ**

GitHubのブランチ名は、開発者がコードを管理する上で不可欠な情報であり、様々な場面で確認することができます。gitコマンドやGitHubのWebインターフェースだけでなく、開発環境やCI/CDパイプラインなど、開発プロセス全体に関わってきます。

**補足**

* **リモートブランチ:** リモートリポジトリにあるブランチは、通常はローカルブランチと区別して表示されます。
* **ブランチの表示方法:** ブランチの表示方法は、使用するGitクライアントやGitHubの設定によって多少異なる場合があります。

**Comment by kagomen** [comment]
https://town.biz/blog/2062#i-4

こんなのもある

> エピックブランチとは「エピック」用に作成したブランチから「タスク」ごとにブランチを作成し、「タスク」が完了するごとに「エピック」用のブランチにマージを行っていく運用方法です。
具体的な例を見てみましょう。大きな機能の中にいくつもの機能があり、その機能を開発していくことを考えてみます。例えばこんな感じです。
ブログ機能（「エピック」チケット番号：PRODUCT-1）
Modelの作成（「タスク」チケット番号：PRODUCT-2）
登録機能（「タスク」チケット番号：PRODUCT-3）
編集機能（「タスク」チケット番号：PRODUCT-4）
一覧機能（「タスク」チケット番号：PRODUCT-5）
詳細機能（「タスク」チケット番号：PRODUCT-6）
削除機能（「タスク」チケット番号：PRODUCT-7）

**Reply by kagomen** [reply]
でも、イシューを細かく切れば良い、というだけの気がする
